### PR TITLE
fix issue with array access code generation

### PR DESCRIPTION
**Pull request by hannasm**
This is intended to resolve issue EamonNerbonne/ExpressionToCode/#50

### CODE REVIEW

File: ExpressionToCodeTest/ArrayAccessTests.cs [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+        Expression GenerateCode<T1,T2>(Expression<Func<T1,T2>> code) {
+            return code;    
+        }
+        [Test]
+        public void Test001()
+        {
+            Assert.AreEqual("a => a[1]",
+                ExpressionToCode.ToCode(
+                GenerateCode<string[],string>(a=>a[1]))
+            );
```
Comment (EamonNerbonne) [inline ExpressionToCodeTest/ArrayAccessTests.cs]: You can avoid the complexity of "GenerateCode" and type paramters by declaring a local variable, e.g. `var a = new[] { "hello", "world" };` and accessing that.

File: ExpressionToCodeTest/ArrayAccessTests.cs [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+        }
+        [Test]
+        public void Test002()
+        {
+            var param = Expression.Parameter(typeof(string).MakeArrayType(), "a");
```
Comment (EamonNerbonne) [inline ExpressionToCodeTest/ArrayAccessTests.cs]: `typeof(string).MakeArrayType()` is equivalent to the more typical `typeof(string[])`

File: ExpressionToCodeTest/ArrayAccessTests.cs [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+namespace ExpressionToCodeTest
+{
+    class ArrayAccessTests
+    {
+        Expression GenerateCode<T1,T2>(Expression<Func<T1,T2>> code) {
+            return code;    
+        }
+        [Test]
+        public void Test001()
```
Comment (EamonNerbonne) [inline ExpressionToCodeTest/ArrayAccessTests.cs]: A better name would be nice: something that emphasizes that this test is for the compiler-generated expression tree as opposed to `Expression.ArrayAccess`.

File: ExpressionToCodeTest/ArrayAccessTests.cs [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+        {
+            Assert.AreEqual("a => a[1]",
+                ExpressionToCode.ToCode(
+                GenerateCode<string[],string>(a=>a[1]))
+            );
+        }
+        [Test]
+        public void Test002()
```
Comment (EamonNerbonne) [inline ExpressionToCodeTest/ArrayAccessTests.cs]: A better name would be nice; one that emphasizes that contrary to what you might expect, the compiler does actually generate an `Expression.ArrayAccess`-like expression for array accesses in a lambda expression.

File: ExpressionToCodeLib/ExpressionToCodeImpl.cs [hunk 5]
```diff
@@ -346,10 +346,23 @@ public void DispatchIndex(Expression e)
...
+                    PrefixOrNull = null
+                });
+            }
+            else
+            {
+                if (ie.Indexer.Name != "Item")
+                {
+                    Sink("." + ie.Indexer.Name); //TODO: is this OK?
+                }
+                args = GetArgumentsForMethod(ie.Indexer.GetIndexParameters(), ie.Arguments);
```
Comment (EamonNerbonne) [inline ExpressionToCodeLib/ExpressionToCodeImpl.cs]: This change is fine.  However, it turns out that named indexers can't be created and aren't likely to ever become relevant.  That means the nested `if` *can* be omitted, which means the surrounding `if` can become a ternary, which means you can use type inference for args.

However, it's a minor point - feel free to leave as is if you prefer.

**Comment by EamonNerbonne** [comment]
Hi @hannasm - did you have a chance to look at my comments yet?

**Comment by hannasm** [comment]
Hi @EamonNerbonne  i think i have made all the changes you were talking about, let me know if there's anything else.

**Comment by EamonNerbonne** [comment]
This is in release 1.5.3.

**Pull request merged**